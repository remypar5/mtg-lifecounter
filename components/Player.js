import React from 'react';
import { StyleSheet, Button, View, Text, TouchableHighlight, LifeButton } from 'react-native';

export default class Player extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            points: props.player.lifeTotal
        };

        this.addLife = this.addLife.bind(this);
        this.subtractLife = this.subtractLife.bind(this);
    }
    
    render() {
        const player = this.props.player;
        const points = this.state.points;
        
        return (
            <View style={ styles.tile }>
                <Text style={ styles.playerName }>{ player.name }</Text>
                <View style={ styles.lifeContainer }>
                    <TouchableHighlight style={ styles.button } 
                        onPress={ () => this.subtractLife(1) } onLongPress={ () => this.subtractLife(10) }>
                        <Text style={ styles.buttonText }>-</Text>
                    </TouchableHighlight>
                    <Text style={ styles.points }>{ points }</Text>
                    <TouchableHighlight style={ styles.button } 
                        onPress={ () => this.addLife(1) } onLongPress={ () => this.addLife(10) }>
                        <Text style={ styles.buttonText }>+</Text>
                    </TouchableHighlight>
                </View>
            </View>
        );
    }

    setLife(points) {
        this.setState({
            points
        });
    }
    
    addLife(points = 1) {
        this.setLife(this.state.points + points);
    }

    subtractLife(points = 1) {
        this.setLife(this.state.points - points);
    }
}

const styles = StyleSheet.create({
    tile: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
        borderColor: 'white',
        borderWidth: 1,
        borderStyle: 'solid',
        width: '100%',
    },
    playerName: {
        color: 'white',
        fontSize: 16,
        textAlign: 'center',
        elevation: 30,
    },
    lifeContainer: {
        elevation: 20,
        width: '100%',
        flex: 1,
        flexDirection: 'row',
        justifyContent: 'center', // Horizontal
        alignItems: 'stretch', // Vertical
    },
    points: {
        width: '20%',
        height: '100%',
        textAlignVertical: 'center',
        alignSelf: 'center',
        textAlign: 'center',
        color: 'white',
        fontSize: 40,
        // backgroundColor: 'red'
    },
    button: {
        width: '40%',
        // backgroundColor: 'steelblue'
    },
    buttonText: {
        fontSize: 20,
        height: '100%',
        color: 'white',
        fontWeight: 'bold',
        textAlign: 'center',
        textAlignVertical: 'center',
        // backgroundColor: 'gray'
    }
});