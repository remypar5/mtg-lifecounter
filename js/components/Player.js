import React from 'react';
import propTypes from 'prop-types';
import { StyleSheet, Text, View } from 'react-native';

import NumberSpinner from './NumberSpinner';
import { FONT_FAMILY, COLOR_FOREGROUND, COLOR_BACKGROUND, COLOR_MARKED } from '../utils/constants';

export default class Player extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            life: props.player.life,
            gameOver: false
        };

        this.onLifeChange = this.onLifeChange.bind(this);
    }

    render() {
        const { color, name } = this.props.player;
        const { gameOver, life } = this.state;
        const { tile, playerName, gameOver: gameOverStyle } = styles;

        return (
            <View style={ [tile, gameOver ? gameOverStyle : null] }>
                <NumberSpinner value={ life }
                    onChange={this.onLifeChange} />
                <Text style={ [playerName, { color }] }>{ name }</Text>
            </View>
        );
    }

    onLifeChange(life) {
        const gameOver = life <= 0;

        this.setState({
            gameOver
        });

        this.props.onGameOver(gameOver);
    }
};

Player.propTypes = {
    player: propTypes.object.isRequired,
    onGameOver: propTypes.func
}

Player.defaultProps = {
    onGameOver: () => {}
}

const styles = StyleSheet.create({
    tile: {
        flex: 1,
        alignItems: 'stretch',
        justifyContent: 'flex-start',
        borderColor: COLOR_BACKGROUND,
        borderWidth: 3,
        borderStyle: 'solid',
        width: '100%',
        height: '100%',
    },
    gameOver: {
        backgroundColor: COLOR_MARKED
    },
    playerName: {
        color: COLOR_FOREGROUND,
        fontSize: 24,
        alignSelf: 'center',
        paddingVertical: 5
    }
});
