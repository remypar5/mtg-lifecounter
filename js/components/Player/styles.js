import { StyleSheet } from 'react-native';
import { COLOR_BACKGROUND, COLOR_MARKED, FONT_SIZE_TINY, FONT_SIZE } from '../../utils/constants';
import { normalizeFontSize } from '../../utils';

const styles = StyleSheet.create({
    tile: {
        flex: 1,
        alignItems: 'stretch',
        justifyContent: 'flex-start',
        borderColor: COLOR_BACKGROUND,
        borderWidth: 3,
        borderStyle: 'solid',
        width: '100%',
        height: '100%',
    },
    gameOver: {
        backgroundColor: COLOR_MARKED,
    },
    playerName: {
        alignSelf: 'center',
        paddingVertical: 5,
    },
    playerNameSmall: {
        fontSize: normalizeFontSize(FONT_SIZE_TINY),
    },
    playerNameLarge: {
        fontSize: normalizeFontSize(FONT_SIZE),
    },
});

export default styles;
