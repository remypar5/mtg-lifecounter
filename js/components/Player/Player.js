import React from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';

import styles from './styles';
import NumberSpinner from '../NumberSpinner';
import Text from '../Text';

export default class Player extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            life: props.player.life,
            gameOver: false,
        };

        this.onLifeChange = this.onLifeChange.bind(this);
    }

    onLifeChange(life) {
        const gameOver = life <= 0;

        this.setState({
            gameOver,
        });

        this.props.onGameOver(gameOver);
    }

    render() {
        const { size, player } = this.props;
        const { color, name } = player;
        const { gameOver, life } = this.state;
        const {
            tile, playerName, playerNameSmall, playerNameLarge, gameOver: gameOverStyle,
        } = styles;
        const playerNameSizeStyle = size === 'small' ? playerNameSmall : playerNameLarge;
        const containerStyle = [tile, gameOver ? gameOverStyle : null];

        return (
            <View style={containerStyle}>
                <NumberSpinner
                    value={life}
                    step={1}
                    stepLarge={10}
                    size={size}
                    onChange={this.onLifeChange}
                />
                <Text
                    type="label"
                    style={[playerName, playerNameSizeStyle, { color }]}
                >
                    {name}
                </Text>
            </View>
        );
    }
}

Player.propTypes = {
    player: PropTypes.object.isRequired,
    size: PropTypes.oneOf(['small', 'large']).isRequired,
    onGameOver: PropTypes.func,
};

Player.defaultProps = {
    onGameOver: () => {},
};
