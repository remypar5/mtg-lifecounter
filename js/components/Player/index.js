import Player from './Player';
import styles from './styles';

export default Player;
export {
    styles,
};
