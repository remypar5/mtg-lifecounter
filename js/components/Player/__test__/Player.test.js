import React from 'react';
import { shallow } from 'enzyme';

import Player from '../Player';

describe('<Player />', () => {
    let component;

    beforeEach(() => {
        component = shallow(<Player player={{ name: 'John Doe', life: 40 }} />);
    });

    it('renders without crashing', () => {
        expect(component).toMatchSnapshot();
    });
});
