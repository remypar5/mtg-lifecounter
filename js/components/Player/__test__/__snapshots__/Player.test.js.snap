// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Player /> renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Player
    onGameOver={[Function]}
    player={
      Object {
        "life": 40,
        "name": "John Doe",
      }
    }
    size="small"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <NumberSpinner
          max={9007199254740991}
          min={-9007199254740991}
          onChange={[Function]}
          size="small"
          step={1}
          stepLarge={10}
          value={40}
        />,
        <Text
          style={
            Object {
              "alignSelf": "center",
              "color": undefined,
              "fontSize": 90,
              "paddingVertical": 5,
            }
          }
          type="label"
        >
          John Doe
        </Text>,
      ],
      "style": Object {
        "alignItems": "stretch",
        "borderColor": "#262626",
        "borderStyle": "solid",
        "borderWidth": 3,
        "flex": 1,
        "height": "100%",
        "justifyContent": "flex-start",
        "width": "100%",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "max": 9007199254740991,
          "min": -9007199254740991,
          "onChange": [Function],
          "size": "small",
          "step": 1,
          "stepLarge": 10,
          "value": 40,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "John Doe",
          "style": Object {
            "alignSelf": "center",
            "color": undefined,
            "fontSize": 90,
            "paddingVertical": 5,
          },
          "type": "label",
        },
        "ref": null,
        "rendered": "John Doe",
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <NumberSpinner
            max={9007199254740991}
            min={-9007199254740991}
            onChange={[Function]}
            size="small"
            step={1}
            stepLarge={10}
            value={40}
          />,
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": undefined,
                "fontSize": 90,
                "paddingVertical": 5,
              }
            }
            type="label"
          >
            John Doe
          </Text>,
        ],
        "style": Object {
          "alignItems": "stretch",
          "borderColor": "#262626",
          "borderStyle": "solid",
          "borderWidth": 3,
          "flex": 1,
          "height": "100%",
          "justifyContent": "flex-start",
          "width": "100%",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "max": 9007199254740991,
            "min": -9007199254740991,
            "onChange": [Function],
            "size": "small",
            "step": 1,
            "stepLarge": 10,
            "value": 40,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "John Doe",
            "style": Object {
              "alignSelf": "center",
              "color": undefined,
              "fontSize": 90,
              "paddingVertical": 5,
            },
            "type": "label",
          },
          "ref": null,
          "rendered": "John Doe",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
